flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskExecutionDecorationTest.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.runtime.tasks;
+
+import org.apache.flink.runtime.checkpoint.CheckpointMetaData;
+import org.apache.flink.runtime.checkpoint.CheckpointOptions;
+import org.apache.flink.runtime.checkpoint.CheckpointType;
+import org.apache.flink.runtime.state.CheckpointStorageLocationReference;
+import org.apache.flink.runtime.util.FatalExitExceptionHandler;
+import org.apache.flink.streaming.api.operators.StreamOperator;
+import org.apache.flink.streaming.runtime.tasks.mailbox.Mail;
+import org.apache.flink.streaming.runtime.tasks.mailbox.MailboxDefaultAction;
+import org.apache.flink.util.function.ThrowingRunnable;
+
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Verifies that {@link StreamTask} {@link StreamTaskActionExecutor decorates execution} of actions that potentially needs to be synchronized.
+ */
+public class StreamTaskExecutionDecorationTest {
+	private CountingStreamTaskActionExecutor decorator;
+	private StreamTask<Object, StreamOperator<Object>> task;
+
+	@Test
+	public void testAbortCheckpointOnBarrierIsDecorated() throws Exception {
+		task.abortCheckpointOnBarrier(1, null);
+		verify();
+	}
+
+	@Test
+	public void testTriggerCheckpointOnBarrierIsDecorated() throws Exception {
+		task.triggerCheckpointOnBarrier(new CheckpointMetaData(1, 2), new CheckpointOptions(CheckpointType.CHECKPOINT, new CheckpointStorageLocationReference(new byte[]{1})), null);
+		verify();
+	}
+
+	@Test
+	public void testTriggerCheckpointAsyncIsDecorated() throws Exception {
```

Comment:
Test setup is quite complicated because of the async call. You also don't test if just the triggering of `triggerCheckpoint` is using the decorator or if it's the actually checkpointing. Since you are also testing if mailbox is using the decorator this test may be redundant.
To simplify and fix the scope, you could make `triggerCheckpoint` visible for testing and use that in this case.

Reply:
The test checks that `actionExecutor` is used in `triggerCheckpointAsync` also outside of `triggerCheckpoint`.

And I'd rather keep private methods private and test only public methods.